File: app/src/main/java/ru/skinallergic/checkskin/components/fooddiary/FoodDiaryFragment.java
package ru.skinallergic.checkskin.components.fooddiary;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.ObservableField;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import ru.skinallergic.checkskin.R;
import ru.skinallergic.checkskin.databinding.FragmentFoodDiaryBinding;
import ru.skinallergic.checkskin.view_models.DateViewModel;
import ru.skinallergic.checkskin.Loger;
import ru.skinallergic.checkskin.components.fooddiary.adapters.RecyclerAdapter_food;
import ru.skinallergic.checkskin.components.fooddiary.data.EntityFoodPosition;
import ru.skinallergic.checkskin.components.fooddiary.viewModels.FoodDiaryViewModel;

import java.util.Date;
import java.util.List;

import static ru.skinallergic.checkskin.components.fooddiary.Util.TIME_OF_MEALS;

public class FoodDiaryFragment extends Fragment {
    private int rawPosition=0;
    private DialogFragment dialogfragment;
    private FoodDiaryViewModel viewModel;
    private RecyclerView recyclerView;
    private MutableLiveData<Integer> rawPositionLive=new MutableLiveData<>();
    private FragmentFoodDiaryBinding binding;
    private String date;
    private DateViewModel dateViewModel;
    public ObservableField<String> dateObservable=new ObservableField<>();

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        Loger.log("• onCreateView");
        viewModel = new ViewModelProvider(requireActivity()).get(FoodDiaryViewModel.class);
        dateViewModel=new ViewModelProvider(requireActivity()).get(DateViewModel.class);
        date=dateViewModel.getDate();
        viewModel.refreshFoodList(date,TIME_OF_MEALS[rawPosition]);

        binding=FragmentFoodDiaryBinding.inflate(inflater);
        binding.setFragment(this);
        View view=binding.getRoot();
        recyclerView=binding.recycler;

        dateObservable.set(dateViewModel.getDate());

        //календарь--------нужно поулчить Long
        //dialogfragment = new DatePickerTheme(dateViewModel.dateLive);

        viewModel.foodLive.observe(getViewLifecycleOwner(),(List<EntityFoodPosition> list)-> {
                RecyclerView.Adapter adapter=new RecyclerAdapter_food(list);
                recyclerView.setLayoutManager(new LinearLayoutManager(getActivity().getBaseContext()));
                recyclerView.setHasFixedSize(false);
                recyclerView.setAdapter(adapter);
                adapter.notifyDataSetChanged();
        });
        dateViewModel.dateLive.observe(getViewLifecycleOwner(),(Date d)-> {
            String date=dateViewModel.getDate(d);
            dateObservable.set(date);
            viewModel.refreshFoodList(date,TIME_OF_MEALS[rawPosition]);
        });
        rawPositionLive.observe(getViewLifecycleOwner(), (Integer raw)-> {
            rawPosition=raw;
            String date=dateViewModel.getDate();
            String timeOfMeals=TIME_OF_MEALS[rawPosition];
            Loger.log("rawPositionLive date: "+date+", timeOfMeals: "+timeOfMeals);
            binding.timeOfMeal.setText(timeOfMeals);
            viewModel.refreshFoodList(date, timeOfMeals);
        });
        return view;
    }
    public void clickDate(View view){
        dialogfragment.show(requireActivity().getSupportFragmentManager(), "theme");
    }
    public void add(View view){
        Navigation.findNavController(view).navigate(R.id.addFoodFragment);
    }
    public void toSearch(View view){
        Navigation.findNavController(view).navigate(R.id.searchFoodFragment);
    }
    public void chooseTimeOfMeal(View view){
        switch (view.getId()){
            case R.id.brackfast:
                rawPositionLive.setValue(0);
                break;
            case R.id.snack:
                rawPositionLive.setValue(1);
                break;
            case R.id.lunch:
                rawPositionLive.setValue(2);
                break;
            case R.id.dinner:
                rawPositionLive.setValue(3);
                break;
        }
        String date=dateViewModel.getDate();
        String timeOfMeals=TIME_OF_MEALS[rawPosition];
        Loger.log("viewModel.dateLive date: "+date+", timeOfMeals: "+timeOfMeals);
        binding.date.setText(date);
        viewModel.refreshFoodList(date,timeOfMeals);
    }

    public void toRedact (View view){
        Loger.log("onClick");
        Bundle bundle=new Bundle();
        bundle.putInt("position", rawPosition);
        bundle.putString("date",dateViewModel.getDate());
        bundle.putString("timeOfMeal",TIME_OF_MEALS[rawPosition]);
        Navigation.findNavController(view).navigate(R.id.redactFoodFragment, bundle);
    }
}

File: app/src/main/java/ru/skinallergic/checkskin/components/fooddiary/components/FoodDiaryFragment.kt
package ru.skinallergic.checkskin.components.fooddiary.components

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import androidx.recyclerview.widget.RecyclerView
import ru.skinallergic.checkskin.Loger
import ru.skinallergic.checkskin.R
import ru.skinallergic.checkskin.components.fooddiary.adapters.FoodPositionAdapter
import ru.skinallergic.checkskin.components.fooddiary.adapters.SwipeRecyclerAdapterFood
import ru.skinallergic.checkskin.components.fooddiary.data.Food
import ru.skinallergic.checkskin.components.fooddiary.data.FoodEntity
import ru.skinallergic.checkskin.components.fooddiary.data.FoodMealForMain
import ru.skinallergic.checkskin.components.fooddiary.viewModels.FoodDiaryViewModel
import ru.skinallergic.checkskin.databinding.FragmentFoodDiary2Binding
import ru.skinallergic.checkskin.databinding.ItemFoodBinding
import ru.skinallergic.checkskin.databinding.ItemOneEatBinding

class FoodDiaryFragment : BaseFoodFragment(), SwipeRecyclerAdapterFood.OnSwipeItemClickListener, SwipeRecyclerAdapterFood.DeleteItemClickListener {
    lateinit var buttonProfile: ImageButton
    lateinit var buttonDate: ImageButton
    lateinit var buttonAllergy: ImageButton
    lateinit var buttonAdd: ViewGroup
    lateinit var recycler: RecyclerView
    lateinit var adapter :SwipeRecyclerAdapterFood
    ////////////////////////////////////////////////// it was "this"  = owner
    val foodDiaryViewModel by lazy { ViewModelProvider(requireActivity(), viewModelFactory).get(FoodDiaryViewModel::class.java) }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val binding =FragmentFoodDiary2Binding.inflate(inflater)
        val view= binding.root
        binding.apply {
            fragment=this@FoodDiaryFragment
            buttonProfile = profileBtn
            buttonDate = dateBtn
            buttonAllergy = allergyBtn
            buttonAdd = addBtn
            recycler=recyclerView
        }

        /*adapter = SwipeRecyclerAdapterFood( parentFragmentManager, arrayListOf()
        ) { binder: ItemFoodBinding, entity: FoodMealForMain ->
            binder.entity = entity
            binder.containerLayout.setOnClickListener { view: View ->
                val bundle = Bundle()
                bundle.putInt(BUNDLE_ID_OF_REMIND, entity.id)
                Loger.log("positionId for bundle " + entity.id)
                navFunction(view, R.id.action_remindersFragment3_to_remindersDetailFragment, bundle)
            }
        }

        adapter .apply {
            setOnItemClickListener(this@FoodDiaryFragment)
            setDeleteItemClickListener(this@FoodDiaryFragment)
            recycler.adapter = this
        }*/

        return view
    }

    override fun onStart() {
        super.onStart()
        subscribeDate()
        buttonProfile.setOnClickListener { Navigation.findNavController(it).navigate(R.id.profileFragment)}
        buttonDate.setOnClickListener { Navigation.findNavController(it).navigate(R.id.action_foodDiaryFragment_to_foodCalendarFragment)}
        buttonAllergy.setOnClickListener { Navigation.findNavController(it).navigate(R.id.action_foodDiaryFragment_to_allergicListFragment)}
        buttonAdd.setOnClickListener { Navigation.findNavController(it).navigate(R.id.action_foodDiaryFragment_to_addFoodFragment2) }

        foodDiaryViewModel.getFoodDiaryByDate(dateViewModel.dateUnix).observe(viewLifecycleOwner, {list->
            for (pos in list){
                println("${pos?.created} ${pos?.list}")
            }

            /*val diffUtil = list?.let { DiffUtilFromMySelf<FoodMealForMain?>(adapter.list, it) }
            diffUtil?.calculate(adapter)*/

            adapter = SwipeRecyclerAdapterFood( parentFragmentManager, list as java.util.ArrayList<FoodMealForMain>?
            ) { binder: ItemFoodBinding, entity: FoodMealForMain ->
                binder.entity = entity
                binder.clickable.setOnClickListener { view: View ->
                    val bundle = Bundle()
                    bundle.putInt(BUNDLE_ID_OF_REMIND, entity.id)
                    Loger.log("positionId for bundle " + entity.id)
                    navFunction(view, R.id.action_foodDiaryFragment_to_detailFoodFragment, bundle)
                }
                val innerRecycler=binder.recycler
                createInnerAdapter(innerRecycler, entity.list as ArrayList<FoodEntity>)
            }

            adapter .apply {
                setOnItemClickListener(this@FoodDiaryFragment)
                setDeleteItemClickListener(this@FoodDiaryFragment)
                recycler.adapter = this
            }

        })
    }

    fun navFunction(view: View, fragmentId: Int, bundle: Bundle){
        Navigation.findNavController(view).navigate(fragmentId, bundle)
    }

    override fun onItemClick(view: View?, id: Int) {
        Loger.log("onItemClick, id $id")
    }

    override fun onItemDelete(id: Int) {
        Loger.log("onItemDelete, id $id")
    }

    fun createInnerAdapter(recyclerView: RecyclerView, list: ArrayList<FoodEntity>){
        val adapter=FoodPositionAdapter(list,object : FoodPositionAdapter.RecyclerCallback<ItemOneEatBinding, FoodEntity>{
            override fun bind(binder: ItemOneEatBinding, food: FoodEntity) {
                binder.food=food
            }
        })
        recyclerView.adapter=adapter
    }
}